File: datastructure/singleylinkedlist.js
class Node {
    constructor(value){
        this.value=value
        this.next=null
    }
 }

 class linkedlist{
    constructor(){
        this.head=null
        this.tail=null
        this.size=0
      
    }


    isempty(){
        return this.size===0
    }


    getsize(){
        return this.size
    }


    piepend(value){
       let node=new Node(value)
        if(this.isempty()){
           this.head=node
           this.tail=node
        }else{
            node.next=this.head
            this.head=node
        }
      this.size++
    }


    insert(index,value){
       if(index<0 || index>this.size){
        return
       }
       if(index===0){
        this.piepend(value)
       }else{
        let node=new Node(value)
        let inc=this.head
        for(let i=0;i<index-1;i++){
            inc=inc.next
        }
        node.next=inc.next
        inc.next=node
       }
       this.size++
    }


    append(value){
        let node=new Node(value)
        if(this.isempty()){
            this.tail=node
            this.head=node
        }else{
            this.tail.next=node
            this.tail=node
        }
        
        this.size++
        
        }

     
    removefromfront(){
        if(this.isempty()){
            return null
        }else{
            let value=this.head.value
            let curr=this.head
            this.head=this.head.next
            curr.next=null
            this.size--
            return value
        }}


    removefromend(){
        if(this.isempty()){
            return null
        }if(this.size===1){
            let value=this.tail.value
            this.tail=null
            this.head=null
            return value
        }else{
            let prev=this.head
            while(prev.next!==this.tail){
                prev=prev.next
            }
            let value=this.tail.value
            prev.next=this.tail.next
            this.tail=prev
            this.size--
            return value
            
        }}


    removeusingindex(index)
    {
        if(index<0 || index>=this.size){
            return null
        }if(index===0){
         let removenode=this.head.value
                     this.head=this.head.next
                     this.size--
                   return removenode 
        }
        else{
            let first=this.head
         
            for(let i=0;i<index-1;i++){
                first=first.next
            }
          
           let last=first.next
                first.next=last.next
           let removenode=last.value
                  last.next=null
                  this.size--
                  return removenode
        }}


    removeusingvalue(value){
        if(this.isempty()){
            return null
        }if(this.head.value===value){
            let curr=this.head
    this.head=this.head.next
    curr.next=null
    return value
        }
        else{
        let search=this.head
       
       for(let i=0;i<this.size;i++){
        if(search.value===value){
                let first=this.head
              
                 for(let j=0;j<i-1;j++){
                    first=first.next
                 }
              
              let  last=first.next
                 first.next=last.next
              let removednode=last.value
                 last.next=null
                 this.size--
                 return removednode.value
        }else{
            search=search.next
            if(i===this.size-1){
                return null

            }
        }}}}


    search(value){
        if(this.isempty()){
            return null

        }
        else{
        let search=this.head
        for(let i=0;i<this.size;i++){
            if(search.value===value){
               return `${value} is found at index:${i}`
            }else{
                search=search.next
                if(i===this.size-1){
                  return  `${value} is not found` 
                }
            }
        }
    }
    }
    reverse(){
     let prev=null
     let curr=this.head
     while(curr){
        let nex=curr.next
        curr.next=prev
        prev=curr
        curr=nex
     }
     this.head=prev
    }

    
    print(){
        if(this.isempty()){
            console.log("no value is stored")
        }
        
        let curr=this.head
        let listvalues=``
        while(curr){
            listvalues+=`${curr.value} `
            curr=curr.next
        }
        console.log(listvalues)
    }
    

        
       
 }

 
 module.exports=linkedlist

File: java/java_datastructure/singleylinkedlist.java
package java_datastructure;

public class singleylinkedlist {
       private ListNode head;
       private ListNode tail;
        private int size;

    public static class  ListNode {
         private int data;
         private ListNode next;

         public ListNode(int value){
            this.data=value;
            this.next=null;
         }      
    }

    public boolean isEmpty(){
        return this.size==0;
    }

    public int getsize(){
        return this.size;
    }

    public void append(int value){
        ListNode node=new ListNode(value);
        if(isEmpty()){
          this.head=node;
          this.tail=node;
        }else{
           this.tail.next=node;
           this.tail=node;
           this.tail.next=null;
        }
        this.size++;
    }

    
     public int getstackelement(){
        if(isEmpty()) return 404;
        ListNode last=tail;
        return last.data;
     }
     public int getqueueelement(){
        if(isEmpty()) return 404;
        ListNode first=head;
        return first.data;
     }
    
    public void piepend(int value){
        ListNode node=new ListNode(value);
        if(isEmpty()){
          this.head=node;
          this.tail=node;
        }else{
            node.next=this.head;
            this.head=node;
        }
        this.size++;
    }

    public void insert(int index,int value){
        if(index<0 || index>this.size){
            return;
        }
        if(index==0){
            piepend(value);
        }else{
            ListNode node =new ListNode(value);
        ListNode curr=head;
        for(int i=1;i<index;i++){
            curr=curr.next;
        }
        node.next=curr.next;
        curr.next=node;
       this.size++; 
        }   
    }

    public int removefromfront(){
        if(isEmpty()){
            return 404;
        }
        ListNode curr=head;
        head=head.next;
        curr.next=null;
        this.size--;
        return curr.data;
    }

    public int removefromindex(int index){
        if(index<0 || index>=this.size){
            return 404;
        }
        if(index==0){
         return removefromfront();
        }
        ListNode curr=head;
        ListNode another=curr.next;
        for(int i=1;i<index;i++){
            curr=curr.next;
            another=another.next;
        }
        curr.next=another.next;
        another.next=null;
        this.size--;
        return another.data;     
    }

    public int removefromlast(){
        if(isEmpty()){
            return 404;
        }
        ListNode curr=head;
        while (curr.next.next!=null){
            curr=curr.next;
        }
        this.tail=curr;
        curr=curr.next;
        tail.next=null;
        this.size--;
        return curr.data;
    }

    public String search(int value){
        if(isEmpty()) return "linkedlist is empty";
        if(this.head.data==value) return value+" is present on index 0";
        ListNode search=head;
        for(int i=0;i<this.size;i++){
            if(search.data==value) return value+" is present on index "+i;
            search=search.next;
        }
        return value+" is not present in linkedlist";
    }

    public void print(){
        ListNode curr=head;
        String result="";
       while(curr!=null){
        result+=curr.data + " ";
        curr=curr.next;
       }
       System.out.println(result);
    }





 public static void main(String[] args) {
singleylinkedlist list=new singleylinkedlist();
   list.append(1);
   list.append(2);
   list.append(3);
   list.piepend(0);
   list.piepend(-1);
   list.insert(5, 4);
   list.print();
   } 

}